song/song_.py

```diff
@@ -107,13 +107,12 @@ def _update_embeddings(self) -> None:
             dire = self.embeddings[i_1] - self.embeddings[j]
             dist = np.linalg.norm(dire)
             e = self.topology[i_1][j]
-            # nabla = 2 * a * b * e * dist ** (2 * b - 2) / (1 + dist ** (2 * b))
-            # TODO above derivative value occurs zero division warning,
-            # so some modification will be needed.
-            nabla = 0.001
+            nabla = (2 * a * b * e * dist ** (2 * b - 2)) / (1 + dist ** (2 * b))
             self.embeddings[j] += self.alpha * nabla * dire
 
         for j in np.random.permutation(list(negative_edges))[:n_negative_sample]:
+            if j == i_1:
```

Comment:
when j is the diagonal component, skip this step to avoid  dire=0 and nabla is NaN